Integrate ground data model objects into client

seanlobo opened the pull request:
`client.py` now returns data model objects instead of the json response from the server when applicable. The `get_*_history` and `get_*_latest_versions` methods are exceptions, as it those return lists/dicts of data instead of objects. Also cleaned up a bit of the style, removed some typos, changed method names from camel case to underscores and modified some default dictionary/lists to be None and initialized in the method.

~This pr doesn't include any tests yet, I can make a separate one when those are finished or add them here later.~ Tests have been added

seanlobo committed:
Implement initial Tag and Item classes

Commit:
Added some models

Commit:
Merge branch 'sean' of github.com:seanlobo/client into andre

Commit:
Reorganized structure

Commit:
Added more strcutre, renamed files

seanlobo committed:
Initial Node class, update import, constructor in Item

Commit:
Added all structure files - need more core objects still

seanlobo committed:
Bug fixes and style changes on existing structs

Fix constructors, add classmethod constructors and various bug fixes /
style changes to most files

seanlobo committed:
Add new objects, fix typo in eq of edge_version

seanlobo committed:
Update imports, syntax, fix typos in datamodel classes

seanlobo committed:
Update client to return data objects, fix typos

seanlobo committed:
Update setup.py to install requests

seanlobo committed:
Rename methods to use underscores

seanlobo committed:
Replace mutable default paramter values with None

seanlobo committed:
A few small bug fixes, typos

seanlobo committed:
Add initial test cases for client

vsreekanti commented on python/ground/client.py:
```diff
@@ -55,54 +58,67 @@ def _getRichVersionJson(self, reference, reference_parameters, tags, structure_v
+                            structure_version_id=-1,
+                            parent_ids=None):
+
+        if reference_parameters is None:
```
Why is this being done here? This should be done in `_get_rich_version_json`.

seanlobo replied:
Looks like in `_get_rich_version_json` the values are only added if they evaluate to `True` anyway, so we can just remove all of these as `None` and empty collections all evaluate to `False`

vsreekanti commented on python/ground/client.py:
```diff
@@ -114,193 +130,284 @@ def createEdgeVersion(self,
-            structure_version_id=-1,
-            parent_ids=[]):
+    def create_graph(self, source_key, name, tags=None):
+        if tags is None:
```
Same here.

vsreekanti replied:
And it seems with all of the `create` methods.

seanlobo replied:
Same as above

vsreekanti commented on python/ground/common/model/core/edge.py:
```diff
@@ -0,0 +1,44 @@
+from ..version.item import Item
+
+
+class Edge(Item):
```
There should also be a constructor that takes in the actual fields. The client methods should be able to take in these objects as arguments to the `create` methods. e.g., `create_edge(edge)`.

seanlobo replied:
Line 4 isn't defining a constructor, it's defining the `Edge` class as extending the `Item` class. On line 13 there's a classmethod `from_edge` which creates a new edge from a previous edge object (usage would be `new_edge = Edge.from_edge(id, old_edge)`. Should that be changed to use `old_edge`'s id (so usage could be `Edge.from_edge(old_edge)`, and they would both share the same id)?

vsreekanti commented on python/ground/common/model/core/structure.py:
```diff
@@ -0,0 +1,37 @@
+from ..version.item import Item
```
Is there no better way to do this? Okay if not, but seems clunky.

seanlobo replied:
Updated these imports to be absolute

vsreekanti reviewed (changes requested):
Mostly looks good! A few minor requests / changes and a couple questions. Will merge once those are addressed.

seanlobo committed:
Remove unnecessary checks on default values

Remove occurences of `if var is None: var = []` from methods when the
variable is only used if it is nonempty

seanlobo committed:
Update imports in datamodel classes to be absolute
